source/request-as-event-emitter.js

```diff
@@ -6,38 +6,56 @@ const http = require('http');
 const https = require('https');
 const urlLib = require('url');
 const CacheableRequest = require('cacheable-request');
+const toReadableStream = require('to-readable-stream');
 const is = require('@sindresorhus/is');
 const timer = require('@szmarczak/http-timer');
-const timedOut = require('./timed-out');
-const getBodySize = require('./get-body-size');
+const timedOut = require('./utils/timed-out');
+const getBodySize = require('./utils/get-body-size');
 const getResponse = require('./get-response');
 const progress = require('./progress');
-const {GotError, CacheError, UnsupportedProtocolError, MaxRedirectsError, RequestError} = require('./errors');
+const {GotError, CacheError, UnsupportedProtocolError, MaxRedirectsError, RequestError, TimeoutError} = require('./errors');
+const urlToOptions = require('./utils/url-to-options');
 
 const getMethodRedirectCodes = new Set([300, 301, 302, 303, 304, 305, 307, 308]);
 const allMethodRedirectCodes = new Set([300, 303, 307, 308]);
 
-module.exports = options => {
+const cachedCacheableRequests = new Map();
+const createCacheableRequest = (request, cache) => {
+	if (!cachedCacheableRequests.has(request)) {
+		cachedCacheableRequests.set(request, new Map());
+	}
+
+	if (!cachedCacheableRequests.get(request).has(cache)) {
+		cachedCacheableRequests.get(request).set(cache, new CacheableRequest(request, cache));
+	}
+
+	return cachedCacheableRequests.get(request).get(cache);
+};
```

Comment:
Won't this cache all requests forever and cause a memory leak?

Reply:
I thought about it too. But this'd require to generate new cache per request.

Perhaps we could limit the size of the cache to 100 items?

Reply:
> Why would it cache the requests?

What do you mean?

> I agree that memoization could be a feature of the package itself or one around it though.

+1

Reply:
Cacheable-request exports a class (for some reason) that actually just gives you the cacheable factory method bound to the actual request factory (`request` in your code, which is a function, right?) and the cache. Calling the return value of `new CacheableRequest` is what actually produces the request, and there's no caching going on at that level. Hence, all you're doing is caching a function and some state, not the request itself. Therefore, v8 should GC it just fine. Correct me if I'm wrong.